src/main/java/uk/nhs/hee/tis/trainee/forms/service/LtftService.java

```diff
@@ -313,26 +318,92 @@ protected Optional<LtftFormDto> changeLtftFormState(UUID formId,
+
+    if (form.isPresent()) {
+      LtftForm updatedForm = updateStatus(form.get(), state, adminIdentity, detail);
+      return Optional.of(mapper.toDto(updatedForm));
+    } else {
+      log.warn("Could not update form {} since no form exists with this ID for DBCs [{}]",
+          formId, dbcs);
       return Optional.empty();
     }
+  }
+
+  /**
+   * Update the status of the LTFT, the current status and history will both be updated.
+   *
+   * @param form        The form to update the status of.
+   * @param targetState The state to change to.
+   * @param identity    Who is performing the status change.
+   * @param detail      A detailed reason for the change, may be null.
+   * @return The updated LTFT application.
+   * @throws MethodArgumentNotValidException If the state transition is not allowed.
+   */
+  private LtftForm updateStatus(LtftForm form, LifecycleState targetState,
+      UserIdentity identity, @Nullable LftfStatusInfoDetailDto detail)
+      throws MethodArgumentNotValidException {
+
+    if (!LifecycleState.canTransitionTo(form, targetState)) {
+      log.warn(
+          "Could not update form {}, invalid lifecycle transition from {} to {} for form type '{}'",
+          form.getId(), form.getStatus().current().state(), targetState, form.getFormType());
+
+      BeanPropertyBindingResult result = new BeanPropertyBindingResult(form, "form");
+      result.addError(new FieldError("LtftForm", "status.current.state",
+          "can not be transitioned to %s".formatted(targetState)));
+      throw new MethodArgumentNotValidException(null, result);
+    }
+
+    if (targetState.isRequiresDetails() && (detail == null || detail.reason() == null)) {
+      log.warn("Form {} requires a reason to change to state [{}]", form.getId(), targetState);
+
+      BeanPropertyBindingResult result = new BeanPropertyBindingResult(detail, "detail");
+      String field = detail == null ? "detail" : "detail.reason";
+      result.addError(new FieldError("StatusInfo", field,
+          "must not be null when transitioning to %s".formatted(targetState)));
+
+      throw new MethodArgumentNotValidException(null, result);
+    }
+
+    StatusDetail detailEntity = mapper.toStatusDetail(detail);
 
     Person modifiedBy = Person.builder()
-        .name(traineeIdentity.getName())
-        .email(traineeIdentity.getEmail())
-        .role(TraineeIdentity.ROLE)
+        .name(identity.getName())
+        .email(identity.getEmail())
+        .role(identity.getRole())
         .build();
-    AbstractAuditedForm.Status.StatusDetail statusDetail = mapper.toStatusDetail(detail);
-    log.info("Changing form {} state to {} for trainee [{}]", formId, targetState, traineeId);
-    form.setLifecycleState(targetState, statusDetail, modifiedBy, form.getRevision());
-    LtftForm savedForm = ltftFormRepository.save(form);
-    return Optional.of(mapper.toDto(savedForm));
+    form.setLifecycleState(targetState, detailEntity, modifiedBy, form.getRevision());
```

Comment:
I'm guessing we can't kick that revision logic much further down the road 😅

Reply:
I think revision and history is my next piece to pick up 👌